### PR TITLE
Adds pen clicking, changes most pen typechecks into writing implement checks

### LINKED ISSUE #84170: Using edagger on closets returns pen behaviors (doesn't allow attacking
## Reproduction:
Get edagger - actuvate it - try attacking a closet

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Fixes #84170 
Adds pen clicking and changes some of edagger and pendriver code to use it instead.
Also replaces most pen typechecks to writing implement checks where it makes sense, so now you can rename things with everything you can write with (crayons)

3d6NwcATNp

Twisting pen caps (for traitor uplinks) has been moved to ctrl + click instead.

## Why It's Good For The Game

Fixes a bug, makes sense that you'd be able to rename stuff with crayons, and clicking is annoying.
Clicky sound included.

## Changelog
:cl:
add: You can now click pens! Clickity clickery click!
fix: E-Daggers can now break lockers/other renamable objects, and pendrivers can now extract their electronics.
code: Most pen typechecks now check for writing implements.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/signs/_signs.dm [hunk 1]
```diff
@@ -187,12 +187,12 @@
 
 /obj/item/sign/add_context(atom/source, list/context, obj/item/held_item, mob/user)
 	. = ..()
-	if(is_editable && istype(held_item, /obj/item/pen))
+	if(is_editable && held_item.get_writing_implement_details()?["interaction_mode"] == MODE_WRITING)
```
Comment (MrMelbert) [inline code/game/objects/structures/signs/_signs.dm:190]: Be careful, you changed some checks like this from something null-safe to something that will runtime

`istype(null, pen)` works, this does not. 

Add context is namely what you should be worried about, attackby and co are safe

File: code/modules/paperwork/pen.dm [hunk 2]
```diff
@@ -212,6 +243,9 @@
 	return TRUE
 
 /obj/item/pen/get_writing_implement_details()
+	var/datum/component/transforming/transform_comp = GetComponent(/datum/component/transforming)
+	if (transform_comp?.active)
+		return null
```
Comment (MrMelbert) [inline code/modules/paperwork/pen.dm:248]: Get component >:(

Just check `TRAIT_TRANSFORM_ACTIVE`

File: code/modules/paperwork/pen.dm [hunk 3]
```diff
@@ -47,6 +49,32 @@
 	AddElement(/datum/element/tool_renaming)
 	RegisterSignal(src, COMSIG_DART_INSERT_ADDED, PROC_REF(on_inserted_into_dart))
 	RegisterSignal(src, COMSIG_DART_INSERT_REMOVED, PROC_REF(on_removed_from_dart))
+	if (!can_click)
+		return
+	RegisterSignal(src, COMSIG_TRANSFORMING_ON_TRANSFORM, PROC_REF(on_transform))
+	if (custom_component)
+		return
```
Comment (MrMelbert) [inline code/modules/paperwork/pen.dm:56]: This var is weird, just have its own proc `setup_transforming` which subtypes can override

File: code/modules/paperwork/pen.dm [hunk 4]
```diff
@@ -34,6 +34,8 @@
 	var/dart_insert_icon = 'icons/obj/weapons/guns/toy.dmi'
 	var/dart_insert_casing_icon_state = "overlay_pen"
 	var/dart_insert_projectile_icon_state = "overlay_pen_proj"
+	var/can_click = TRUE // If this pen can be clicked in order to retract it
```
Comment (MrMelbert) [inline code/modules/paperwork/pen.dm:37]: Autodoc this (`///` above the comment)

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Maybe we should have a helper macro for checking if something is a writing utensil

`#define IS_WRITING_UTENSIL(thing) (thing.get_writing_implement_details()?["interaction_mode"] == MODE_WRITING)`?

**Comment by optimumtact** [comment]
you should make it so if someone clicks conseuctively 100,000 times without being interrupted they get an achievement + they get greentext (and become valid)

**Comment by LSommers** [comment]
> you should make it so if someone clicks consecutively 100,000 times without being interrupted they get an achievement + they get greentext (and become valid)

Every click above 100 in a row should give an ever-higher chance of making your arm blow off from the stress, but make the 100,000 clicks a per-round counter.

Make them earn that achievement.

**Comment by SmArtKar** [comment]
> you cant open pen uplinks with this change

you can, been moved to ctrl + click functionality

**Pull request merged**